### PR TITLE
Introduces TypeScript

### PR DESCRIPTION
**Pull request by KaiVolland**
### Description:

This transforms the project to use and be build with TypeScript.

Componentss:

- [x] Button
- [x] CircleMenu
- [x] Container
- [x] Field
- [x] Grid
- [x] HigherOrderComponent
- [x] LayerTree
- [x] LayerTreeNode
- [x] Legend
- [x] Map
- [x] Panel
- [x] Provider
- [x] Slider
- [x] Toolbar
- [x] UserChip
- [x] Window
- [x] Util

TODOs:

- [x] Tests
- [x] Styleguide
  - [x] Fix Examples
 

Closes #1240
Closes #1241

### CODE REVIEW

**Comment by KaiVolland** [comment]
tsc currently failing because of: https://github.com/ag-grid/ag-grid/issues/3492

**Comment by KaiVolland** [comment]
The build is currently failing due to a bug in ag-grid/ag-grid#3492 when using the latest TypeScript version (3.7.2). Therefore the TypeScript version is pinned to "~3.6.0". I'll open an issue once this is PR is merged.

File: package.json [hunk 1]
```diff
@@ -101,22 +108,23 @@
     "@babel/polyfill": "^7.4.4",
     "@babel/preset-env": "^7.5.5",
     "@babel/preset-react": "^7.0.0",
+    "@babel/preset-typescript": "^7.3.3",
+    "@types/enzyme": "^3.10.3",
+    "@types/jest": "^24.0.22",
+    "@types/react": "^16.9.11",
```
Comment (dnlkoch) [inline package.json]: Please move to `dependencies`.

File: webpack.common.config.js [hunk 2]
```diff
@@ -1,15 +1,43 @@
...
     https: true
   },
-
+  resolve: {
+    extensions: [
+      '.ts',
+      '.tsx',
+      '.js',
+      '.json',
+      '.jsx',
```
Comment (dnlkoch) [inline webpack.common.config.js]: Do we still have any `*.jsx` files to handle? If not, I suggest to remove support for these files completely.

File: babel.config.js [hunk 3]
```diff
@@ -1,25 +1,14 @@
...
       ]
-    },
-    build: {
-      plugins: [
-        [
-          'module-resolver',
-          {
-            resolvePath(sourcePath) {
-              return sourcePath.replace('.jsx', '.js');
-            }
```
[the remark this answers is not in the document]
Comment (KaiVolland) [reply babel.config.js]: As we produce the production build with TypeScript now the babel.config.js is only used for Jest. So I don't think this is relevant anymore.

File: package.json (hunk 1 above)
Comment (KaiVolland) [reply package.json]: We had several issues when using "@types/react" as depency in a library. So i think we should keep it as devDependency and peerDependency.

File: package.json [hunk 4]
```diff
@@ -53,22 +53,22 @@
     "clean:build": "rimraf ./build/*",
     "clean:dist": "rimraf ./dist/*",
     "pretest": "npm run lint",
-    "test": "jest --maxWorkers=4",
+    "test": "jest --maxWorkers=4 --coverage",
     "test:watch": "jest --watchAll",
     "coveralls": "cat coverage/lcov.info | coveralls",
     "prepublishOnly": "npm run build",
-    "lint:styleguide": "eslint -c .eslint-styleguide.js --ext md src/",
```
[the remark this answers is not in the document]
Comment (KaiVolland) [reply package.json]: I don't think so. We could try to readd eslint just for the examples if it is important to you.

[the remark this answers is not in the document]
Comment (KaiVolland) [reply package.json]: ... I readded the eslint stuff to lint the examples.

File: src/Button/ToggleGroup/ToggleGroup.tsx [hunk 5]
```diff
@@ -1,76 +1,65 @@
-import React from 'react';
-import PropTypes from 'prop-types';
+import * as React from 'react';
+import * as PropTypes from 'prop-types';
```
Comment (ahennr) [inline src/Button/ToggleGroup/ToggleGroup.tsx]: See above comment.

File: src/Button/UploadButton/UploadButton.tsx [hunk 6]
```diff
@@ -18,40 +34,14 @@ import './UploadButton.less';
  * @class The UploadButton
  * @extends React.Component
  */
-class UploadButton extends React.Component {
+class UploadButton extends React.Component<UploadButtonProps> {
```
Comment (ahennr) [inline src/Button/UploadButton/UploadButton.tsx]: Would it make sense to include an empty state interface here as well for the sake of completeness?

File: src/Grid/PropertyGrid/PropertyGrid.tsx [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    feature: OlFeature,
+    attributeFilter: string[],
+    attributeNames: AttributeNames,
+    attributeNameColumnWidthInPercent: number
+  }): {
+    dataSource: any,
+    columns: ColumnProps<any>[]
+  } {
+    if (!attributeFilter) {
+      attributeFilter = feature.getKeys().filter((attrName) => attrName !== 'geometry');
```
Comment (ahennr) [inline src/Grid/PropertyGrid/PropertyGrid.tsx]: Is the data type missing here?

File: src/Grid/PropertyGrid/PropertyGrid.tsx [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+    }
+
+    const dataSource = attributeFilter.map((attr) => {
```
Comment (ahennr) [inline src/Grid/PropertyGrid/PropertyGrid.tsx]: Data type inherited from `attributeFilter`?

File: src/Button/ToggleButton/ToggleButton.tsx [hunk 9]
```diff
@@ -1,63 +1,94 @@
-import React from 'react';
-import PropTypes from 'prop-types';
+import * as React from 'react';
+import * as PropTypes from 'prop-types';
```
[the remark this answers is not in the document]
Comment (KaiVolland) [reply src/Button/ToggleButton/ToggleButton.tsx]: The React context doesn't work without `static contextTypes`. So i had to keep the PropTypes for some components.

File: src/Button/UploadButton/UploadButton.tsx (hunk 6 above)
Comment (KaiVolland) [reply src/Button/UploadButton/UploadButton.tsx]: No. I don't think we should do that.

File: src/Button/DigitizeButton/DigitizeButton.tsx [hunk 10]
```diff
@@ -13,30 +12,213 @@ import OlStyleStroke from 'ol/style/Stroke';
...
+  /**
+   * Whether the line, point, polygon, circle, rectangle or text shape should
+   * be drawn.
+   */
+  drawType?: DrawType;
+  /**
+   * Whether the digitize feature should be deleted, copied or modified.
+   * be drawn.
+   */
+  editType?: 'Copy' | 'Edit' | 'Delete';
```
[the remark this answers is not in the document]
Comment (KaiVolland) [reply src/Button/DigitizeButton/DigitizeButton.tsx]: This Pros **extend** the `Partial<DigitizeButtonDefaultProps>` so i think you have to define again what is optional.

**Comment by KaiVolland** [comment]
Thx for the review @ahennr. There is still a lot room for TypeScript improvements and you already found some lines.
This PR is jus the introduction and we should improve typings as much as possible in the near future.

File: src/Button/UploadButton/UploadButton.tsx (hunk 6 above)
Comment (ahennr) [reply src/Button/UploadButton/UploadButton.tsx]: Alright.

**Comment by KaiVolland** [comment]
Bug in ag-grid types was fixed in ag-grid v 22.0.0 so i could upgrade to TypeScript 3.7.2.

**Comment by KaiVolland** [comment]
Thx for the reviews.

**Comment by marcjansen** [comment]
Very nice friends!

**Pull request merged**